### PR TITLE
Flax: Fix img2img and align with other pipeline

### PR DESCRIPTION
**Pull request by skirsten**
- Fixes for flax img2img
- Other misc changes
- Re-aligned the img2img pipe with the normal pipe (mostly copy paste)

### CODE REVIEW

File: src/diffusers/pipeline_flax_utils.py
Comment (pcuenca) [inline src/diffusers/pipeline_flax_utils.py:506]:
```suggestion
            A dictionary containing all the modules needed to initialize the pipeline.
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:403]:
```suggestion
    pipe,
```

Can we please remove the line comments here?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:409]:
```suggestion
    num_inference_steps,
    height,
    width,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:425]:
```suggestion
    pipe,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:433]:
```suggestion
    start_timestep,
    num_inference_steps,
    height,
    width,
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Ideally, we could make it so that `strength` can be a sharded array too (behaving exactly like `guidance_scale`) so we can perform multiple parallel requests with different strengths. This would be useful for batching requests, but we can address it in a future PR.

**Review by pcuenca — COMMENTED** [review]
Thanks a lot! This clearly improves the existing pipeline, I just left comments with a few suggestions and ideas for a potential improvement (make `strength` parallelizable too).

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Initially I also wanted to do that, but AFIK its not possible (in an easy way):

The strength influences the `start_timestep` which is used in the `jax.lax.fori_loop` bounds. The bounds for the loop must be known at compile time (as with `num_inference_steps`) so AFAIK there is no way to have a different strength per shard in pmap.

The only way I could think of doing that is passing in a static `min_start_timestep` and using that as the loop bounds and then also passing in the dynamic sharded `start_timestep`. Then in the loop body, using a jax conditional discard / "skip" the body if `timestemp < start_timestep` but that would be overkill in my opinion.

Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Btw, the reason I did the `strength` -> `start_timestep` calculation outside the jit is to reduce the number of recompilations.
E.g. `strength(0.8) != strength(0.81)` but `start_timestep(0.8) == start_timestep(0.81)`

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:169]: why is `offset` removed? We need to handle this for `DDIMScheduler`

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Agree with @skirsten, `strength` is used to determine the value of inference steps, and different `strength = different number of timesteps`, so it would be hard to parallelize it.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:223]: any specific reason to hardcode the `dtype` to `jnp.float32`? think we should use `self.dtype` here as before for half-precision inference.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:273]: think it's safer to use `len(scheduler_state.timesteps)` that `num_inference_steps` because depending on the scheduler the could be some extra timesteps, for example if using `PNDMScheduler` with PRK stpes, it'll add some extra time timesteps

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:193]: we compute the `height` and `width` in `__call__` as well, so here we could make it required.

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for working on this! Left some comments

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:182]: Yes, good points @skirsten and @patil-suraj, let's leave it like this.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:169]: The offset cancels itself out. It was also removed from the other img2img pipelines in #1653

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:223]: This was a copy paste from the normal pipeline. There are some other places where the `dtype` is hardcoded to `jnp.float32` related to latents. I just remember setting every occurrence to `self.dtype` and losing all details in the generated images.

I would prefer if all occurrences of `jnp.float32` could be removed at the same time. But I can also just revert this one change. Let me know what you think.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:273]: :thinking: I was not aware of that. Thanks for letting me know. In that case it would also have to be changed [here](https://github.com/huggingface/diffusers/blob/df2b548e893ccb8a888467c2508756680df22821/src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py#L271).

Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:273]: Hmm, it looks to me like [here](https://github.com/huggingface/diffusers/blob/25e82ec47c40d5619e0772ed68f4082da4bd6457/src/diffusers/schedulers/scheduling_pndm_flax.py#L200) it is making sure that the length is always equal to `num_inference_steps` by dropping some plms timesteps.

Or am I getting that wrong?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:169]: Oh I see, didn't notice that, thanks!

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:219]: cc @pcuenca do we need noise in float32 in Jax?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: Actually you are right, let's use `num_inference_steps` here. Because the timesteps are extended for 2nd order schedulers like Heun.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (skirsten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:219]: I investigated this a bit more and found the "problem". I asked the JAX maintainers on their thoughts here: https://github.com/google/jax/discussions/13798

Basically, to prevent losing detail in the image, the noise has to be generated in `float32` and then casted to `self.dtype`:
```py
noise = jax.random.normal(prng_seed, shape=latents_shape, dtype=jnp.float32).astype(self.dtype)
```

Still I would prefer not to do that in this PR and instead create a new PR that fixes all of these occurrences at once.

**Review by patil-suraj — APPROVED** [review]
All looks good now, thank you for addressing the comments! The quality checks are failing, run `make style` and `make quality` , push and then it should be good to merge after green light from @pcuenca

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:219]: That experiment was very cool and instructive, thanks a lot for taking the time to clarify the behaviour! I agree to deal with this in another PR.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: Hmm, I'm not sure about this. Because the timesteps are extended for some schedulers, shouldn't we loop through the `timesteps` instead?

Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: Those schedulers are not part of flax yet no? (specifically the `HeunScheduler`),

Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_img2img.py:269]: No, they are not.

**Review by pcuenca — APPROVED** [review]

**Comment by skirsten** [comment]
The failing CI seems to be unrelated to these changes (doc_builder)

**Comment by pcuenca** [comment]
@skirsten This is what `make style` did for me: a47d1f07, you probably didn't install `doc-builder`.

Feel free to cherry-pick in your branch :)

**Comment by skirsten** [comment]
@pcuenca Thanks, I added you commit. Yes, I was missing some packages (and make did not resolve python to python3) so I just assumed that the problem was not in this branch :see_no_evil:

**Comment by pcuenca** [comment]
No worries!

**Comment by pcuenca** [comment]
@skirsten @patil-suraj is this ready to merge then?

**Pull request merged**
